File: backend-go/internal/httpapi/respond.go
package httpapi

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"

	"github.com/danielraab/na-notes-app/backend-go/internal/apperr"
)

func respondJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

func respondNoContent(w http.ResponseWriter) {
	w.WriteHeader(http.StatusNoContent)
}

type errorBody struct {
	Error struct {
		Code    string `json:"code"`
		Message string `json:"message"`
	} `json:"error"`
}

func respondError(w http.ResponseWriter, status int, code, message string) {
	var body errorBody
	body.Error.Code = code
	body.Error.Message = message
	respondJSON(w, status, body)
}

// respondDomainError maps the small set of sentinel errors domain
// packages return onto HTTP status codes. Anything else is treated as an
// unexpected server error and logged with detail the client never sees.
func respondDomainError(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, apperr.ErrNotFound):
		respondError(w, http.StatusNotFound, "NOT_FOUND", "resource not found")
	case errors.Is(err, apperr.ErrForbidden):
		respondError(w, http.StatusForbidden, "FORBIDDEN", "not permitted")
	case errors.Is(err, apperr.ErrValidation):
		respondError(w, http.StatusBadRequest, "VALIDATION_ERROR", err.Error())
	case errors.Is(err, apperr.ErrVersionConflict):
		respondError(w, http.StatusConflict, "VERSION_CONFLICT", "note was modified since you last loaded it")
	default:
		slog.Error("unhandled error", "error", err)
		respondError(w, http.StatusInternalServerError, "INTERNAL", "internal server error")
	}
}

func decodeJSON(r *http.Request, v any) error {
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	return dec.Decode(v)
}

File: backend-rust/src/httpapi/respond.rs
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::json;

use crate::apperr::AppError;

/// The single error response shape (`{"error": {"code", "message"}}`)
/// every endpoint uses, and the small mapping from `apperr::AppError` onto
/// HTTP status codes. Anything unexpected is logged with detail the client
/// never sees.
pub struct ApiError {
    pub status: StatusCode,
    pub code: &'static str,
    pub message: String,
}

impl ApiError {
    pub fn new(status: StatusCode, code: &'static str, message: impl Into<String>) -> ApiError {
        ApiError {
            status,
            code,
            message: message.into(),
        }
    }

    pub fn unauthenticated() -> ApiError {
        ApiError::new(
            StatusCode::UNAUTHORIZED,
            "UNAUTHENTICATED",
            "login required",
        )
    }
}

impl IntoResponse for ApiError {
    fn into_response(self) -> Response {
        let body = json!({ "error": { "code": self.code, "message": self.message } });
        (self.status, Json(body)).into_response()
    }
}

impl From<AppError> for ApiError {
    fn from(err: AppError) -> Self {
        match err {
            AppError::NotFound => {
                ApiError::new(StatusCode::NOT_FOUND, "NOT_FOUND", "resource not found")
            }
            AppError::Forbidden => {
                ApiError::new(StatusCode::FORBIDDEN, "FORBIDDEN", "not permitted")
            }
            AppError::Validation(msg) => {
                ApiError::new(StatusCode::BAD_REQUEST, "VALIDATION_ERROR", msg)
            }
            AppError::Internal(msg) => {
                tracing::error!(error = %msg, "unhandled error");
                ApiError::new(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "INTERNAL",
                    "internal server error",
                )
            }
        }
    }
}
